plugins/externalwrench/src/ExternalWrench.cc

```diff
@@ -161,13 +161,122 @@ void ExternalWrench::applyWrench()
+        ignition::math::Vector3d torque (wrench.torque[0], wrench.torque[1], wrench.torque[2]);
+
+        //Frames:
+        //World frame = W
+        //Link frame = L
+        //Link CoG frame = Lg
+        //Applied Forece frame = F
+
+        //Quantities:
+        //Rotation Matrix = R
+        //Transformation Matrix = T
+        //Quaternion = Q
+        //Pose 3d = P
+        //Position 3d = p
+
+        //Examples:
+        //W_T_L = the Transformation Matrix from the World frame to the Link frame
+        //Lg_Q_F = the Quaternion from the Link CoG frame to the Applied Force frame
+
+        ignition::math::Vector3d linkCoGPos = link->WorldCoGPose().Pos(); // Get link's COG position where wrench will be applied
+        ignition::math::Vector3d newZ = force.Normalized(); // normalized force. I want the z axis of the cylinder's reference frame to coincide with my force vector
+        ignition::math::Vector3d newX = newZ.Cross (ignition::math::Vector3d::UnitZ);
+        ignition::math::Vector3d newY = newZ.Cross (newX);
+
+        ignition::math::Matrix4d Lg_T_F = ignition::math::Matrix4d (newX[0],newY[0],newZ[0],0,newX[1],newY[1],newZ[1],0,newX[2],newY[2],newZ[2],0, 0, 0, 0, 1);
+        ignition::math::Quaterniond Lg_Q_F = Lg_T_F.Rotation();
+
+        ignition::math::Vector3d W_p_Lg = link->WorldCoGPose().Pos(); // Get link's COG position where wrench will be applied
+        ignition::math::Quaterniond W_Q_Lg = link->WorldCoGPose().Rot();
+        ignition::math::Matrix4d W_T_Lg = ignition::math::Matrix4d (link->WorldCoGPose());
+
+        link->AddRelativeForce(force);
+        link->AddRelativeTorque(torque);
+
+        ignition::math::Quaterniond W_Q_F = W_Q_Lg*Lg_Q_F;
+        ignition::math::Matrix4d W_T_F = W_T_Lg*Lg_T_F;
+
+        const ignition::math::Vector3d cylinderHalfLength = ignition::math::Vector3d ( 0,0,-0.15 );
+
+        ignition::math::Pose3d W_p_F (W_T_F*cylinderHalfLength, W_Q_F);
+#else
+        math::Vector3d force (wrench.force[0], wrench.force[1], wrench.force[2]);
+        math::Vector3d torque (wrench.torque[0], wrench.torque[1], wrench.torque[2]);
+
+        //Frames:
+        //World frame = W
+        //Link frame = L
+        //Link CoG frame = Lg
+        //Applied Forece frame = F
+
+        //Quantities:
+        //Rotation Matrix = R
+        //Transformation Matrix = T
+        //Quaternion = Q
+        //Pose 3d = P
+        //Position 3d = p
+
+        //Examples:
+        //W_T_L = the Transformation Matrix from the World frame to the Link frame
+        //Lg_Q_F = the Quaternion from the Link CoG frame to the Applied Force frame
+
+        math::Vector3d linkCoGPos = link->WorldCoGPose().Pos(); // Get link's COG position where wrench will be applied
+        math::Vector3d newZ = force.Normalized(); // normalized force. I want the z axis of the cylinder's reference frame to coincide with my force vector
+        math::Vector3d newX = newZ.Cross (math::Vector3d::UnitZ);
```

Comment:
Should we consider here the UnitZ of the Gazebo world ? or the UnitZ of the link?

Reply:
I think the current setting is OK (in Gazebo world). Since UnitZ is used to construct the transformation from the applied force frame to the Link CoG frame (`Lg_T_F`), which then is multiplied to the transformation between the world frame and the Link CoG frame:
```
        ignition::math::Matrix4d W_T_F = W_T_Lg*Lg_T_F;
```
In [this line](https://github.com/HosameldinMohamed/gazebo-yarp-plugins/blob/c00ff7eaf8ec8d23cf01f18092ee5d5270da783c/plugins/externalwrench/src/ExternalWrench.cc#L235) for example.